# OpenDartReader - 비상장 기업 데이터 조회와 활용

<img width="40%" src="https://i.imgur.com/NE65l5L.png" >

#### 2019-2020 [FinanceData.KR](http://financedata.kr) | [facebook.com/financedata](http://facebook.com/financedata)

## 요약

```python
import OpenDartReader

api_key = '961aebb762ef44e8f93a36218897eb**********'
dart = OpenDartReader(api_key)

# [비바리퍼블리카] (송금서비스 토스로 유명한), 공시(감사보고서) 목록
df = dart.list('비바리퍼블리카', start='1900') 

# [빅히트엔터테인먼트] (BTS로 유명한), 공시(감사보고서) 목록
df = dart.list('빅히트엔터테인먼트', start='1900')
df = dart.list('01204056', start='1900')

# [쿠팡] "연결감사보고서 (2019.12)" 20200414000766 (삼일회계법인)
# http://dart.fss.or.kr/dsaf001/main.do?rcpNo=20200414000766
html_text = dart.document('20200414000766')
df_list = pd.read_html(html_text)
df_list[15] # 연결포괄손익계산서
```

<img src="https://i.imgur.com/c3t2toM.png" >


## OpenDartReader 설치와 초기화

In [1]:
!pip install -q OpenDartReader # 1회만 설치

In [2]:
import OpenDartReader

# 발급받은 API키를 사용하세요
api_key = '961aebb762ef44e8f93a36218897eb**********'

dart = OpenDartReader(api_key)

## 비상장 회사의 공시보고서 목록

In [3]:
df = dart.list('비바리퍼블리카', start='1900') # 송금서비스 토스
df

corp_code corp_name stock_code corp_cls  ...        rcept_no  flr_nm  rcept_dt rm
0  01212921   비바리퍼블리카                   E  ...  20200410003129  삼일회계법인  20200410   
1  01212921   비바리퍼블리카                   E  ...  20190412002776  삼일회계법인  20190412   
2  01212921   비바리퍼블리카                   E  ...  20180413001411  삼일회계법인  20180413   

[3 rows x 9 columns]

In [4]:
# 보고서 링크
for ix, row in df.iterrows():
    rcp_no = row['rcept_no']
    print(f'http://dart.fss.or.kr/dsaf001/main.do?rcpNo={rcp_no}')

http://dart.fss.or.kr/dsaf001/main.do?rcpNo=20200410003129
http://dart.fss.or.kr/dsaf001/main.do?rcpNo=20190412002776
http://dart.fss.or.kr/dsaf001/main.do?rcpNo=20180413001411


## 이름으로 확인 하기

회사 이름으로 식별이 고유하게 식별이 가능한지 확인해야 하는 것을 추천합니다.

예를 들어, BTS로 유명한 '빅히트엔터테인먼트'를 조회하면 정확히 1개 회사가 있지만, '빅히트'로 검색하면 '빅히트'가 포함된 여러 회사가 있을 수 있습니다.

In [5]:
comps = dart.company_by_name('빅히트엔터테인먼트')
len(comps)

1

In [6]:
comps = dart.company_by_name('빅히트')
len(comps)

4

In [7]:
import pandas as pd
pd.DataFrame(comps)

status message corp_code  ... induty_code    est_dt acc_mt
0    000      정상  00294858  ...       64201  20000107     01
1    000      정상  01204056  ...       73901  20050201     12
2    000      정상  01449288  ...       46452  20191001     12
3    000      정상  01449367  ...       90191  20191001     12

[4 rows x 19 columns]

## 공시목록 가져오기

조회를 해보고 정식명칭의 전체 회사명을 사용하십시오. `corp_code`를 쓰는 것이 가장 확실합니다.

In [8]:
bighit = dart.list('빅히트엔터테인먼트', start='1900')
bighit

corp_code  corp_name stock_code  ...     flr_nm  rcept_dt rm
0  01204056  빅히트엔터테인먼트             ...  빅히트엔터테인먼트  20200902   
1  01204056  빅히트엔터테인먼트             ...     한영회계법인  20200331   
2  01204056  빅히트엔터테인먼트             ...     한영회계법인  20200331   
3  01204056  빅히트엔터테인먼트             ...     삼일회계법인  20190319   
4  01204056  빅히트엔터테인먼트             ...     삼일회계법인  20180322   

[5 rows x 9 columns]

In [9]:
bighit = dart.list('01204056', start='1900')
bighit

corp_code  corp_name stock_code  ...     flr_nm  rcept_dt rm
0  01204056  빅히트엔터테인먼트             ...  빅히트엔터테인먼트  20200902   
1  01204056  빅히트엔터테인먼트             ...     한영회계법인  20200331   
2  01204056  빅히트엔터테인먼트             ...     한영회계법인  20200331   
3  01204056  빅히트엔터테인먼트             ...     삼일회계법인  20190319   
4  01204056  빅히트엔터테인먼트             ...     삼일회계법인  20180322   

[5 rows x 9 columns]

## 공시 본문가져오기

"연결감사보고서 (2019.12)"의 접수번호(rcept_no) '20200331000038'로 공시보고서 문서 전체를 얻을 수 있습니다. (형식은 XHTML 입니다)

http://dart.fss.or.kr/dsaf001/main.do?rcpNo=20200331000038




In [10]:
html_text = dart.document('20200331000038')
html_text[:1000] # 앞부분

'<?xml version="1.0" encoding="utf-8"?>\n\n\n<DOCUMENT xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="dart3.xsd">\n\n<DOCUMENT-HEADER AEXT-CLASS="Y">\n<DOCUMENT-NAME ACODE="00761">연결감사보고서</DOCUMENT-NAME>\n<FORMULA-VERSION SUBVER="4" ADATE="20190212">3.4</FORMULA-VERSION>\n<COMPANY-NAME AREGCIK="01204056" AACCOUNTTYPE="A">주식회사 빅히트엔터테인먼트</COMPANY-NAME>\n\n<SUMMARY>\n<EXTRACTION ACODE="AUDIT_CIK" AFEATURE="BOTH">00260198</EXTRACTION>\n<EXTRACTION ACODE="SUPV_OPIN" AFEATURE="BOTH">100000000000</EXTRACTION>\n<EXTRACTION ACODE="TOT_ASSETS" AFEATURE="BOTH">362990</EXTRACTION>\n<EXTRACTION ACODE="TOT_DEBTS" AFEATURE="BOTH">189467</EXTRACTION>\n<EXTRACTION ACODE="TOT_SALES" AFEATURE="BOTH">587224</EXTRACTION>\n<EXTRACTION ACODE="TOT_EMPL" AFEATURE="BOTH">228</EXTRACTION>\n<EXTRACTION ACODE="ETC_OTHR_YN" AFEATURE="BOTH">Y</EXTRACTION>\n<EXTRACTION ACODE="POW_OTHR_YN" AFEATURE="BOTH">N</EXTRACTION>\n<EXTRACTION ACODE="ACC_PROF_YN" AFEATURE="BOTH">Y</EXTRACTIO

## 재무데이터를 데이터프레임으로 읽어오기

pandas의 read_html()로 페이지에 포함된 표를 손쉽게 DataFrame으로 읽을 수 있습니다.

In [11]:
df_list = pd.read_html(html_text)

In [12]:
len(df_list)

142

페이지안에 총 142개의 테이블(table)이 있습니다.

몇 번째 표가 `연결재무상태표`인지는 인덱스번호를 넣어가며 찾아야 합니다.  여기서는 15번째(인덱스로는 14)가 연결재무상태표 입니다.

In [13]:
# 연결재무상태표
df = df_list[14]

remove_text = "&cr;(감사받지 아니한 재무제표)"
df.columns = [col.replace(remove_text, '') for col in df.columns]
df

과 목         주 석  ...     제14(전) 기초   제14(전) 기초.1
0                  자산         NaN  ...           NaN           NaN
1                유동자산         NaN  ...           NaN      55866023
2            현금및현금성자산        33.0  ...      29504093           NaN
3   당기손익-공정가치 측정 금융자산     43233.0  ...      12004106           NaN
4                매출채권     53033.0  ...       8208888           NaN
5            기타유동금융자산       633.0  ...       4999066           NaN
6              기타유동자산         7.0  ...       1115220           NaN
7                재고자산         8.0  ...         34650           NaN
8               비유동자산         NaN  ...           NaN      10148197
9   당기손익-공정가치 측정 금융자산     43233.0  ...       1585539           NaN
10          기타비유동금융자산       633.0  ...       4479710           NaN
11             공동기업투자         9.0  ...             -           NaN
12               유형자산        10.0  ...       1497060           NaN
13              사용권자산        34.0  ...       1756495           NaN
14               무형자산        11.0  ...        239583           NaN
15            이연법인세자산        18.0  ...        353020           NaN
16            기타비유동자산         7.0  ...        236790           NaN
17              자산 총계         NaN  ...           NaN      66014220
18                 부채         NaN  ...           NaN           NaN
19               유동부채         NaN  ...           NaN      93981466
20               매입채무      3133.0  ...             -           NaN
21          전환상환우선주부채    143133.0  ...       5856344           NaN
22               리스부채    313334.0  ...       1029084           NaN
23           기타유동금융부채  12303133.0  ...      78151121           NaN
24            당기법인세부채         NaN  ...       4479798           NaN
25             기타유동부채        13.0  ...       4465119           NaN
26            유동성장기부채        15.0  ...             -           NaN
27               충당부채        16.0  ...             -           NaN
28              비유동부채         NaN  ...           NaN      13997870
29          전환상환우선주부채    143133.0  ...      11479292           NaN
30               리스부채    313334.0  ...        161326           NaN
31              장기차입금        15.0  ...             -           NaN
32          기타비유동금융부채    123033.0  ...        159377           NaN
33            순확정급여부채        17.0  ...       1963663           NaN
34            이연법인세부채        18.0  ...             -           NaN
35            기타비유동부채        13.0  ...        167138           NaN
36               충당부채        16.0  ...         67074           NaN
37              부채 총계         NaN  ...           NaN     107979336
38                 자본         NaN  ...           NaN           NaN
39         지배기업의 소유지분         NaN  ...           NaN  (41,965,116)
40                자본금       119.0  ...        525026           NaN
41            주식발행초과금        19.0  ...        265179           NaN
42               기타자본        20.0  ...       2335460           NaN
43          기타포괄손익누계액        20.0  ...      (13,775)           NaN
44                결손금         NaN  ...  (45,077,006)           NaN
45              비지배지분         NaN  ...           NaN             -
46              자본 총계         NaN  ...           NaN  (41,965,116)
47          부채와 자본 총계         NaN  ...           NaN      66014220

[48 rows x 8 columns]

In [14]:
# 연결포괄손익계산서
df = df_list[18]

remove_text = "&cr;(감사받지 아니한 재무제표)"
df.columns = [col.replace(remove_text, '') for col in df.columns]
df

과 목     주 석  ...       제14기        제14기.1
0                        매출액  2230.0  ...        NaN     301371824
1                       매출원가  2430.0  ...        NaN     189978918
2                      매출총이익     NaN  ...        NaN     111392906
3                    판매비와관리비  2324.0  ...   31462308           NaN
4                       영업이익     NaN  ...        NaN      79930598
5                       기타수익    25.0  ...      97511           NaN
6                       기타비용    25.0  ...    1768914           NaN
7                      지분법손실     9.0  ...     124381           NaN
8                       금융수익  2633.0  ...    2207145           NaN
9                       금융비용  2633.0  ...  128791219           NaN
10               법인세비용차감전순손익     NaN  ...        NaN  (48,449,260)
11                     법인세비용    18.0  ...        NaN      22016319
12                     당기순손익     NaN  ...        NaN  (70,465,579)
13        지배기업의 소유주 귀속 당기순이익     NaN  ...        NaN  (70,465,579)
14            비지배지분 귀속 당기순이익     NaN  ...        NaN             -
15            법인세비용차감후기타포괄손익     NaN  ...        NaN      (54,113)
16           당기손익으로 재분류되는 항목     NaN  ...        NaN        149134
17                  해외사업환산차이     NaN  ...     149134           NaN
18  후속적으로 당기손익으로 재분류되지 않는 항목     NaN  ...        NaN     (203,247)
19            순확정급여부채의 재측정요소     NaN  ...  (203,247)           NaN
20                     총포괄손익     NaN  ...        NaN  (70,519,692)
21        지배기업의 소유주 귀속 총포괄손익     NaN  ...        NaN  (70,519,692)
22            비지배지분 귀속 총포괄손익     NaN  ...        NaN             -
23              주당순이익 (단위:원)     NaN  ...        NaN           NaN
24                   기본주당순손익    27.0  ...        NaN      (54,030)
25                   희석주당순손익    27.0  ...        NaN      (54,030)

[26 rows x 6 columns]

In [15]:
# 연결현금흐름표
df = df_list[24]

remove_text = "&cr;(감사받지 아니한 재무제표)"
df.columns = [col.replace(remove_text, '') for col in df.columns]
df

구 분   주 석  ...          제14기        제14기.1
0          영업활동으로 인한 현금흐름   NaN  ...           NaN      74686749
1         영업으로부터 창출된 현금흐름  28.0  ...      88169367           NaN
2                  이자의 수취   NaN  ...         89461           NaN
3                  이자의 지급   NaN  ...         (117)           NaN
4                 법인세의 납부   NaN  ...  (13,571,962)           NaN
5          투자활동으로 인한 현금흐름   NaN  ...           NaN  (45,068,033)
6              기타금융자산의 감소   NaN  ...      71656309           NaN
7           기타비유동금융자산의 감소   NaN  ...         86064           NaN
8                유형자산의 처분   NaN  ...             -           NaN
9       종속기업 취득에 따른 순현금흐름   NaN  ...             -           NaN
10      관계기업 및 공동기업투자의 취득   NaN  ...   (3,360,000)           NaN
11             기타금융자산의 증가   NaN  ...  (95,971,454)           NaN
12          기타비유동금융자산의 증가   NaN  ...   (3,426,566)           NaN
13               유형자산의 취득   NaN  ...   (6,609,511)           NaN
14               무형자산의 취득   NaN  ...   (7,165,982)           NaN
15            기타비유동자산의 증가   NaN  ...     (276,893)           NaN
16         재무활동으로 인한 현금흐름   NaN  ...           NaN   (1,393,794)
17              비지배지분의 증가   NaN  ...             -           NaN
18              주식선택권의 행사   NaN  ...             -           NaN
19                차입금의 차입   NaN  ...        342650           NaN
20                차입금의 상환   NaN  ...     (342,650)           NaN
21             기타금융부채의 감소   NaN  ...      (24,000)           NaN
22               리스부채의 감소   NaN  ...   (1,369,794)           NaN
23           현금및현금성자산의 증가   NaN  ...           NaN      28224922
24  외화표시 현금및현금성자산의 환율변동효과   NaN  ...           NaN        335723
25            기초 현금및현금성자산   NaN  ...           NaN      29504093
26            기말 현금및현금성자산   NaN  ...           NaN      58064738

[27 rows x 6 columns]

## 쿠팡 손익계산서

In [16]:
coupang = dart.list('쿠팡', start='1900')
coupang

corp_code corp_name stock_code corp_cls  ...        rcept_no  flr_nm  rcept_dt rm
0  01019166        쿠팡                   E  ...  20200414000766  삼일회계법인  20200414   
1  01019166        쿠팡                   E  ...  20200414000763  삼일회계법인  20200414   
2  01019166        쿠팡                   E  ...  20190415000123  삼일회계법인  20190415   
3  01019166        쿠팡                   E  ...  20190415000121  삼일회계법인  20190415   
4  01019166        쿠팡                   E  ...  20180416000061  삼일회계법인  20180416   
5  01019166        쿠팡                   E  ...  20180416000060  삼일회계법인  20180416   
6  01019166        쿠팡                   E  ...  20170414001794  삼일회계법인  20170414   
7  01019166        쿠팡                   E  ...  20170414001793  삼일회계법인  20170414   
8  01019166        쿠팡                   E  ...  20160414001921  삼일회계법인  20160414   
9  01019166        쿠팡                   E  ...  20150414002045  삼일회계법인  20150414   

[10 rows x 9 columns]

In [17]:
# http://dart.fss.or.kr/dsaf001/main.do?rcpNo=20200414000766
html_text = dart.document('20200414000766')

df_list = pd.read_html(html_text)
len(df_list)

119

In [18]:
# 연결포괄손익계산서
df_list[15]

과 목        주석   제 7 (당)기     제 6 (전)기
0               Ⅰ. 매출액  182324.0    7153094      4354586
1                 매출원가    1924.0    5950017      4142781
2             영업 일반관리비  192324.0    1923606      1339769
3              Ⅱ. 영업손실       NaN  (720,529)  (1,127,964)
4                 기타수익       NaN       1782          436
5                 기타비용       NaN       1481          666
6                 금융수익     520.0      50396        19338
7                 금융비용     520.0      54155        33450
8       Ⅲ. 법인세비용차감전순손실       NaN  (723,987)  (1,142,306)
9             Ⅳ. 법인세비용      14.0      (747)         2331
10            Ⅴ. 당기순손실       NaN  (723,240)  (1,144,637)
11          지배기업 소유주지분       NaN  (723,240)  (1,144,637)
12               비지배지분       NaN          -            -
13           Ⅵ. 기타포괄손실       NaN   (10,504)      (2,826)
14  당기손익으로 재분류되지 않는 항목       NaN   (10,504)      (2,826)
15      순확정급여부채의 재측정요소      13.0   (10,504)      (2,826)
16          Ⅶ. 당기총포괄손실       NaN  (733,744)  (1,147,463)
17          지배기업 소유주지분       NaN  (733,744)  (1,147,463)
18               비지배지분       NaN          -            -

당기(2019년)에 영업손실 -7,205 억원을 기록했습니다. 전기(2018년)에 -1조1279억원에 비하면 많은 개선이 이루어졌습니다.


#### 2019-2020 [FinanceData.KR](http://financedata.kr) | [facebook.com/financedata](http://facebook.com/financedata)